File: go/pkg/statemachine/statemachine.go
// Package statemachine provides a simple state machine implementation
package statemachine

import (
	"fmt"
)

// StateKey represents a unique key for a state in the state machine.
type StateKey string

// ActionFunc is called when a state is executed.
type ActionFunc[Model any, Input any] func(model *Model, input Input) (key StateKey, err error)

// String returns the string representation of the state key.
func (k StateKey) String() string {
	return (string)(k)
}

// State represents a state in the state machine with a key, name, and action.
type State[Model any, Input any] struct {
	Key    StateKey
	action ActionFunc[Model, Input]
}

// String returns the string representation of the state.
func (s *State[Model, Input]) String() string {
	return fmt.Sprintf("key: %s", s.Key)
}

// GetKey returns the key of the state.
func (s *State[Model, Input]) GetKey() StateKey {
	return s.Key
}

// Execute performs the state's action and returns the key of the next state.
func (s *State[Model, Input]) Execute(model *Model, input Input) (key StateKey, err error) {
	return s.action(model, input)
}

// NewState creates a new state with the given key and action.
func NewState[Model any, Input any](
	key StateKey,
	act ActionFunc[Model, Input],
) *State[Model, Input] {
	if act == nil {
		panic("action function cannot be nil")
	}
	return &State[Model, Input]{
		Key:    key,
		action: act,
	}
}

// ============================================================================

// StateMachine represents a state machine with a current state and a collection of states.
type StateMachine[Model any, Input any] struct {
	currentState *State[Model, Input]
	states       map[StateKey]*State[Model, Input]
	name         string
}

// NewStateMachine creates a new state machine with the given model and name.
func NewStateMachine[Model any, Input any](model *Model, name string) *StateMachine[Model, Input] {
	return &StateMachine[Model, Input]{
		currentState: nil,
		states:       make(map[StateKey]*State[Model, Input]),
		name:         name,
	}
}

// String returns the string representation of the state machine.
func (sm *StateMachine[Model, Input]) String() string {
	return fmt.Sprintf("name: %s , states: %d", sm.name, len(sm.states))
}

// GetCurrentState returns the current state of the state machine.
func (sm *StateMachine[Model, Input]) GetCurrentState() *State[Model, Input] {
	return sm.currentState
}

// GetStates returns a map of all available states.
func (sm *StateMachine[Model, Input]) GetStates() map[StateKey]*State[Model, Input] {
	states := make(map[StateKey]*State[Model, Input])
	for k, v := range sm.states {
		states[k] = v
	}
	return states
}

// AddState adds a new state to the state machine. If it's the first state, it sets it as the initial state.
func (sm *StateMachine[Model, Input]) AddState(state *State[Model, Input]) error {
	// check if state already exists
	if _, exists := sm.states[state.Key]; exists {
		return fmt.Errorf("state %v already exists", state.Key)
	}
	// ok, add it to the map
	sm.states[state.Key] = state

	// if there is no current state, set it as the initial state
	if sm.currentState == nil {
		if err := sm.SetInitialState(state.Key); err != nil {
			return err
		}
	}
	return nil
}

// SetInitialState sets the initial state of the state machine using the given key.
func (sm *StateMachine[Model, Input]) SetInitialState(key StateKey) error {

	state, exists := sm.states[key]
	if !exists {
		return fmt.Errorf("state %v does not exist", key)
	}

	sm.currentState = state

	return nil
}

// Execute performs the current state's action and transitions to the next state based on the returned key.
func (sm *StateMachine[Model, Input]) Execute(model *Model, input Input) (key StateKey, err error) {
	if sm.currentState == nil {
		return "", fmt.Errorf("no current state set")
	}

	key, err = sm.currentState.Execute(model, input)
	if err != nil {
		return
	}

	// same state, no change
	if key == sm.currentState.GetKey() {
		return key, nil
	}

	// new state
	newState, exists := sm.states[key]
	if !exists {
		return "", fmt.Errorf("state %v does not exist", key)
	}

	// set next state
	sm.currentState = newState

	return newState.GetKey(), nil
}

File: c/src/state_machine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "state_machine.h"

#define INITIAL_CAPACITY 10

StateMachine* new_state_machine(const char* name) {
    StateMachine* sm = malloc(sizeof(StateMachine));
    if (!sm) return NULL;

    sm->states = malloc(sizeof(State*) * INITIAL_CAPACITY);
    if (!sm->states) {
        free(sm);
        return NULL;
    }

    sm->current_state = NULL;
    sm->num_states = 0;
    sm->capacity = INITIAL_CAPACITY;
    sm->name = strdup(name);
    
    return sm;
}

void free_state_machine(StateMachine* sm) {
    if (!sm) return;
    
    for (size_t i = 0; i < sm->num_states; i++) {
        free_state(sm->states[i]);
    }
    
    free(sm->states);
    free((void*)sm->name);
    free(sm);
}

State* new_state(StateKey key, ActionFunc action, void* data) {
    if (!key || !action) return NULL;
    
    State* state = malloc(sizeof(State));
    if (!state) return NULL;
    
    state->key = strdup(key);
    state->action = action;
    state->data = data;
    
    return state;
}

void free_state(State* state) {
    if (!state) return;
    free((void*)state->key);
    free(state);
}

static State* find_state(const StateMachine* sm, StateKey key) {
    for (size_t i = 0; i < sm->num_states; i++) {
        if (strcmp(sm->states[i]->key, key) == 0) {
            return sm->states[i];
        }
    }
    return NULL;
}

int add_state(StateMachine* sm, State* state) {
    if (!sm || !state) return -1;
    
    if (find_state(sm, state->key)) {
        return -1;  // State already exists
    }
    
    if (sm->num_states >= sm->capacity) {
        size_t new_capacity = sm->capacity * 2;
        State** new_states = realloc(sm->states, sizeof(State*) * new_capacity);
        if (!new_states) return -1;
        
        sm->states = new_states;
        sm->capacity = new_capacity;
    }
    
    sm->states[sm->num_states++] = state;
    
    if (!sm->current_state) {
        sm->current_state = state;
    }
    
    return 0;
}

int set_initial_state(StateMachine* sm, StateKey key) {
    if (!sm) return -1;
    
    State* state = find_state(sm, key);
    if (!state) return -1;
    
    sm->current_state = state;
    return 0;
}

StateKey execute(StateMachine* sm, Model* model, const void* input) {
    if (!sm || !sm->current_state) return NULL;
    
    StateKey next_key = sm->current_state->action(sm->current_state, model, input);
    if (!next_key) return NULL;
    
    if (strcmp(next_key, sm->current_state->key) != 0) {
        State* next_state = find_state(sm, next_key);
        if (!next_state) return NULL;
        sm->current_state = next_state;
    }
    
    return sm->current_state->key;
}

State* get_current_state(const StateMachine* sm) {
    return sm ? sm->current_state : NULL;
}

const char* state_machine_to_string(const StateMachine* sm, char* buffer, size_t size) {
    if (!sm || !buffer) return NULL;
    snprintf(buffer, size, "name: %s , states: %zu", sm->name, sm->num_states);
    return buffer;
}

const char* state_to_string(const State* state, char* buffer, size_t size) {
    if (!state || !buffer) return NULL;
    snprintf(buffer, size, "key: %s", state->key);
    return buffer;
}
